File: 20221118-Cache/java/app/src/main/java/Cache.java
import java.util.*;

// 이해
// 캐시 사이즈는 하나의 값을 담을 수 있는 크기
// 캐시 사이즈에 새로 들어오는 값이 있다면 그것을 참조해서 cache hit 1
// 없다면 가장 오랫동안 쓰이지 않은 값을 바꾸고 cache miss 5
// 풀이
// List 컨테이너를 준비한다.
// 원하는 값이 담겨 있다면 그 값을 제거하고 새로 밀어 넣는다.
// 없다면 그냥 밀어 넣는다.
// 마지막 인덱스가 head
// 0인덱스가 tail 이 되도록한다.
// jeju seoul busan
// -> seoul busan jeju ( count += 5 )
public class Cache {
  public int solution(int cacheSize, String[] cities) {
    List<String> cacheContainer = new ArrayList<>(cacheSize);
    int count = 0;
    for(String city : cities) {
        city = city.toLowerCase();
     if(cacheContainer.contains(city)) {
       count +=1;
       cacheContainer.remove(city);
       cacheContainer.add(city);
       System.out.println("container: " +  Arrays.toString(new List[]{cacheContainer}));
     } else {
       count +=5;
       if(cacheContainer.size() < cacheSize) {
         cacheContainer.add(city);
         System.out.println("container: " +  Arrays.toString(new List[]{cacheContainer}));
       } else {
         cacheContainer.add(city);
         cacheContainer.remove(0);
         System.out.println("container: " +  Arrays.toString(new List[]{cacheContainer}));
       }
     }
    }

    int answer = count;

    return answer;
  }
}

File: 20221118-Cache/javaScript/Cache.js
// export function remove(array, value) {
//   const indexArray = [];
//   array.forEach((element, index) => (
//     element === value ? indexArray.push(index)
//       : indexArray.push(-1)
//   ));

//   array.splice(indexArray[0], 1);
// }

// export default function solution(cacheSize, cities) {
//   const count = [];
//   const container = [];
//   const unitedCities = cities.map((city) => (
//     city.toLowerCase()
//   ));

//   unitedCities.forEach((city) => {
//     if (cacheSize === 0) {
//       return count.push(5);
//     }
//     if (container.length < cacheSize) {
//       container.push(city);

//       return count.push(5);
//     }
//     if (container.length === cacheSize && container.includes(city)) {
//       remove(container, city);

//       container.push(city);

//       return count.push(1);
//     }
//     if (container.length === cacheSize && !container.includes(city)) {
//       container.splice(0, 1);
//       container.push(city);
//       return count.push(5);
//     }
//     return null;
//   });

//   const answer = count.reduce((sum, currentValue) => (sum + currentValue), 0);
//   return answer;
// }

export default function solution(cacheSize, cities) {
  const map = new Map();
  const cacheHit = (city, map) => {
    map.delete(city);
    map.set(city, city);
    return 1;
  };

  const cacheMiss = (city, map, size) => {
    if (size === 0) return 5;
    (map.size === size) && map.delete(map.keys().next().value);
    map.set(city, city);
    return 5;
  };

  const getTimeCache = (city, map, size) => (map.has(city.toLocaleLowerCase())
    ? cacheHit : cacheMiss)(city.toLocaleLowerCase(), map, size);

  return cities.map(
    (city) => getTimeCache(city.toLocaleLowerCase(), map, cacheSize),
  ).reduce((a, c) => a + c, 0);
}
